packages/graph/hash_graph/lib/interval-ops/src/bounds.rs

```diff
@@ -0,0 +1,261 @@
+    }
+
+    fn into_bound(self) -> Bound<T> {
+        self.map_or(Bound::Unbounded, Bound::Included)
+    }
+
+    fn from_bound(bound: Bound<T>) -> Self {
+        match bound {
+            Bound::Included(value) => Some(value),
+            Bound::Excluded(_) => {
+                panic!("excluded lower bounds are not supported on canonical intervals")
+            }
+            Bound::Unbounded => None,
+        }
+    }
+}
+
+impl<T> UpperBound<T> for Option<T> {
+    fn as_bound(&self) -> Bound<&T> {
+        self.as_ref().map_or(Bound::Unbounded, Bound::Excluded)
+    }
+
+    fn into_bound(self) -> Bound<T> {
+        self.map_or(Bound::Unbounded, Bound::Excluded)
+    }
+
+    fn from_bound(bound: Bound<T>) -> Self {
+        match bound {
+            Bound::Included(_) => {
+                panic!("included upper bounds are not supported on canonical intervals")
+            }
+            Bound::Excluded(value) => Some(value),
+            Bound::Unbounded => None,
+        }
+    }
+}
+
+pub trait LowerBoundComparison<T: PartialOrd>: LowerBound<T> {
+    fn cmp_lower(&self, other: &impl LowerBound<T>) -> Ordering {
+        compare_bounds(
+            self.as_bound(),
+            other.as_bound(),
+            BoundType::Lower,
+            BoundType::Lower,
+        )
+    }
+
+    fn cmp_upper(&self, other: &impl UpperBound<T>) -> Ordering {
+        compare_bounds(
+            self.as_bound(),
+            other.as_bound(),
+            BoundType::Lower,
+            BoundType::Upper,
+        )
+    }
+
+    #[cfg(any(feature = "canonical", feature = "continuous"))]
+    fn cmp_values(&self, other: &impl UpperBound<T>) -> Ordering {
+        compare_bound_values(
+            self.as_bound(),
+            other.as_bound(),
+            BoundType::Lower,
+            BoundType::Upper,
+        )
+    }
+}
+
+impl<B, T> LowerBoundComparison<T> for B
+where
+    T: PartialOrd,
+    B: LowerBound<T>,
+{
+}
+
+pub trait UpperBoundComparison<T: PartialEq>: UpperBound<T> {
+    fn cmp_lower(&self, other: &impl LowerBound<T>) -> Ordering
+    where
+        T: PartialOrd,
+    {
+        compare_bounds(
+            self.as_bound(),
+            other.as_bound(),
+            BoundType::Upper,
+            BoundType::Lower,
+        )
+    }
+
+    fn cmp_upper(&self, other: &impl UpperBound<T>) -> Ordering
+    where
+        T: PartialOrd,
+    {
+        compare_bounds(
+            self.as_bound(),
+            other.as_bound(),
+            BoundType::Upper,
+            BoundType::Upper,
+        )
+    }
+
+    fn is_adjacent_to(&self, other: &impl LowerBound<T>) -> bool {
+        match (self.as_bound(), other.as_bound()) {
+            (Bound::Included(lhs), Bound::Excluded(rhs))
+            | (Bound::Excluded(lhs), Bound::Included(rhs)) => lhs == rhs,
+            _ => false,
+        }
+    }
+}
+
+impl<B, T> UpperBoundComparison<T> for B
+where
+    T: PartialOrd,
+    B: UpperBound<T>,
+{
+}
```

Comment:
Yes. The trait is used to make the implementation details private while keeping the nicer way to call the methods instead of having free-standing functions.